features/api-layer/src/main/java/org/opennms/features/apilayer/collectors/ServiceCollectorManager.java

```diff
@@ -0,0 +1,69 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2019 The OpenNMS Group, Inc.
+ * OpenNMS(R) is Copyright (C) 1999-2019 The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is a registered trademark of The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.features.apilayer.collectors;
+
+import java.util.Hashtable;
+import java.util.Map;
+
+import org.opennms.features.apilayer.utils.InterfaceMapper;
+import org.opennms.integration.api.v1.collectors.ServiceCollectorFactory;
+import org.opennms.netmgt.collection.api.ServiceCollector;
+import org.osgi.framework.BundleContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ServiceCollectorManager extends InterfaceMapper<ServiceCollectorFactory, ServiceCollector> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ServiceCollectorManager.class);
+
+    public ServiceCollectorManager(BundleContext bundleContext) {
+        super(ServiceCollector.class, bundleContext);
+    }
+
+    @Override
+    public ServiceCollector map(ServiceCollectorFactory ext) {
+        return new ServiceCollectorImpl(ext);
+    }
+
+    // Need to override as registry needs collector class name in properties.
+    public synchronized void onBind(ServiceCollectorFactory extension, Map properties) {
+        LOG.debug("bind called with {}: {}", extension, properties);
+        if (extension != null) {
+            extServiceRegistrationMap.computeIfAbsent(extension, (ext) -> {
+                final ServiceCollector mappedExt = map(ext);
+                final Hashtable<String, Object> props = new Hashtable<>();
+                // Make the service available to any Spring-based listeners
+                props.put("registration.export", Boolean.TRUE.toString());
+                // Registry needs type of collector class.
+                props.put("type", ext.getCollectorClassName());
```

Comment:
We should keep the onBind logic in the InterfaceMapper and only extend what we need. Instead of overriding onBind completely I would add something like:
```
    @Override
    public Map<String, Object> getServiceProperties(ServicePollerFactory extension) {
        return ImmutableMap.<String,Object>builder()
                // Registry needs type of collector class.
                .put("type", extension.getCollectorClassName())
                .build();
    }
```

Reply:
We can also refactor the `ServicePollerManager` to do the same.